### PR TITLE
[1.0] WIP Upgrade React Router to v4

### LINKED ISSUE #937: [1.0] Upgrade to React Router v4
https://www.meetfabric.com/ is built on Gatsby 0.x. I'm helping them currently upgrade to Gatsby v1. As part of this work, they're integrating their app (currently a standalone React site) and their marketing website so they'll all be powered by Gatsby.

This is a pattern I'd really love to see more of since Gatsby v1's automatic code splitting makes this cheap to do. Land on a marketing page, login, and immediately jump into the app with no jarring reload. Use the same tools and components and styles across all your web properties!

But as they'd built their app using React Router v4, we decided it'd be best if Gatsby v1 got updated to RRv4 as well (currently we're still on RRv2).

I'll be working on this the next few days and it'll get rolled out as part of the next alpha.

Here's a rough list for the TODOs:

- [x] since RR v4 doesn't need statically defined routes, out routes info as a JSON object and dynamically construct routes in client.
- [x] restore 404 support
- [x] refactor code splits
  - [x] all gatsby code in commons.js
  - [x] all layout and page components have their own code splits
  - [x] extract common code from layout components in addition to page.
- [x] server rendering of routes
- [ ] Inline routes information in each HTML page. Defer loading full routes until after JS initialized.
- [ ] use `window.caches` for caching linked pages https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/caches
- [ ] reevaluate which JS bundles gatsby-plugin-offline precaches
- [ ] let layouts have graphql queries
- [ ] let layouts have parent layouts

### PR DESCRIPTION
**Pull request by KyleAMathews**
Fixes #937 

Current status, rendering things correctly with the development server.

screen shot 2017-05-08 at 5 34 55 pm

### CODE REVIEW

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit 9f7f0805a7bbe12f6dea66e87458cbb1492763cf

https://app.netlify.com/sites/gatsbyjs/deploys/59110f856f4c503f59218854

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit 9f7f0805a7bbe12f6dea66e87458cbb1492763cf

https://app.netlify.com/sites/gatsbygram/deploys/59110f856f4c503f59218856

**Comment by MoOx** [comment]
Curious to understand how do you plan to handle recursive routes. Any detail to share?

**Comment by KyleAMathews** [comment]
@MoOx every page component is at the heart of things and it have 1 or 0 layout components. Generally sites have 1 to N page components and often just 1 layout component (`layouts/index.js` is the default layout component if a page doesn't specify one specifically).

Up to now v1 of Gatsby hasn't supported multiple levels of layouts (which v0 does) but in this PR I'm actually adding back support for that. Basically a layout will be able to indicate that it has a parent component. So when Gatsby creates its routes it'll just recurse up the layouts to the root component and then render that hierarchy.

**Comment by MoOx** [comment]
Oh so Gatsby generate the routes for react-router... I was thinking user were able define their own routes :) So what the point of using react-router?

**Comment by KyleAMathews** [comment]
Something's gotta handle routes and components.

Also RR v4 makes it far easier to embed client-only routes within a larger static Gatsby site. meetfabric.io, the company sponsoring the upgrade, is doing exactly that as they'll be using Gatsby for both their marketing website as well as the client app portion of things.

**Comment by KyleAMathews** [comment]
It's possible at some point we could make RR an optional dependency to slim things further but not today.

File: packages/gatsby/lib/query-runner/route-writer.js [hunk 1]
```diff
@@ -13,6 +13,87 @@ import {
 // Loop through all paths and write them out to child-routes.js
 const writeChildRoutes = () => {
   const { program, config, pages } = store.getState()
+
+  // console.log(pages)
+  // Write out routes.json
+  const routesData = _.merge(
```
Comment (fabien0102) [inline packages/gatsby/lib/query-runner/route-writer.js]: Refactor proposition:

```javascript
const routesData = pages.reduce((mem, {path, componentChunkName, layout, jsonName}) => (
      { ...mem, ...{[path]: {componentChunkName, layout, jsonName}} }
    ), {})
```

https://codepen.io/fabien0102/pen/MmrqKe?editors=0011#

Comment (KyleAMathews) [inline packages/gatsby/lib/query-runner/route-writer.js]: I like it! I'll make that change. Object/array spreading makes reducing quite nice!

**Comment by KyleAMathews** [comment]
This PR has two breaking changes.

* Gatsby now looks for a HTML element with an ID of `___gatsby` to render itself instead of the current `react-mount`. You'll need to change that in your `html.js`. Gatsby may or may not stay React only in the future after all ;-)
* The "default" layout component should now be at `layouts/index.js` not `layouts/default.js`. A quick move over will fix things.

**Comment by KyleAMathews** [comment]
Ok, this looks ready to go in.

One more breaking change — this.props.children in layout components is now a "function". I spent hours trying to upgrade `gatsbygram` yesterday before realizing this was the only option basically.  The problem is that when you load Gatsbygram and click on an image, it loads a modal to display the image but leaves the background the same. So basically the gatsbygram layout needs to say, "keep rendering `/` despite us having navigated now to `/path-to-image/`.

If `this.props.children` is a function, then this is easy. See the code diff for all the gory details https://github.com/gatsbyjs/gatsby/pull/940/files#diff-abbddb44eecd9dea4fc2d307be2d7b9dR157

See this article for a good intro to children as a function https://medium.com/merrickchristensen/function-as-child-components-5f3920a9ace9

This is an easy change though. Just change `this.props.children` in your layout file to `this.props.children()`.

**Comment by KyleAMathews** [comment]
This PR also adds support for client-only routes (and an example site showing this off).

Basically how it works is that a page can have a `matchPath` in addition to its `path`. The path is what's used for server rendering but in the client, `matchPath` will be used in addition to `path` for deciding which page component is in control.

So say you had a marketing website and client-only app in the same Gatsby site.

Marketing pages would be at paths like `/`, `/blog/`, `/pricing/`, `/features/`, etc. The app might be at `/app/`. But the app also has a bunch of sub pages like `/app/settings/` and so on. But these shouldn't be rendered on the server, just on the client. But without modifying anything, when you try to navigate to `/app/settings/`, things will break because Gatsby will try to find that page and fail. So what you do is add a `matchPath` to the `app` page like `/app/:path`. Now for any subpath of `/app/` Gatsby will still pass control of the site to the app page.

The easiest way to add a `matchPath` currently is in `gatsby-node.js` using the API `onUpsertPage` which is called whenever a page is added.

You'll also need to make sure that whenever a user loads from the server urls like `/app/settings/` that you load `/app/index.html`.

**Comment by KyleAMathews** [comment]
I had some tasks for this upgrade that I'm not going to get to right now. Multiple layouts + layout hierarchies, and layouts w/ graphql queries. This is very doable just ran out of time + the API could be tricky so it deserves a deeper dive + discussion.

**Comment by KyleAMathews** [comment]
Create an issue if you're interested in working on it and I'll write up some thoughts.

**Comment by KyleAMathews** [comment]
I snuck in some optimizations. `gatsby-plugin-google-analytics` dropped `react-ga` as a dependency as it had a ton of code we don't need. `gatsby-plugin-catch-links` dropped a dependency and inlined a simple version that does exactly what we need.

**Comment by KyleAMathews** [comment]
This is the URL of the example site I built for client-only routes https://client-only-paths.gatsbyjs.org/

**Pull request merged**